aptos-move/block-executor/src/scheduler.rs

```diff
@@ -240,7 +261,18 @@ impl Scheduler {
             stored_deps.push(txn_idx);
         }
 
-        Some(dep_condvar)
+        // When parallel execution halts, done is set first, and then threads waiting on dependencies are notified.
+        // By checking done here, we ensure this isn't a condvar due to stale state in the multi-version
+        // data-structure that will never be notified. If halt() had already passed the index of the
+        // current thread when scheduler.wait_for_dependency set the status, then done was set even earlier,
+        // and it will be observed. Otherwise, the status will be observed by the terminating thread and
+        // condvar will be notified.
+        if self.done() {
```

Comment:
another way of doing it is to set all txn_status to halted during resolve_condvar, and checked the status in the suspend function so the mutex would be the natural synchronization point.

Reply:
This is what we are doing, but only this is not enough. We need to halt the Move VM for threads that are pending on the read (which may create new condvar). We can have suspend function return something to tell the VM to halt, but it is the same as adding the check here.

Reply:
currently we only set `TransactionStatus::Suspended(_, condvar) | TransactionStatus::ReadyToExecute(_, Some(condvar))` to halted during resolve_condvar, what I meant is to set every txn status to halted so that any subsequent suspend would just fail

Reply:
IIUC, what you are proposing is to check txn_status to synchronize on ExecutionHalted? I don't see the advantage over using done_marker for synchronization as the current code.